src/Microsoft.AspNet.Http.Abstractions/Authentication/AuthenticationManager.cs

```diff
@@ -11,41 +14,93 @@ public abstract class AuthenticationManager
+        // Leave it up to authentication handler to do the right thing for the challenge
+        public Task ChallengeAsync(string authenticationScheme, AuthenticationProperties properties)
         {
-            Challenge(authenticationScheme: null, properties: properties);
+            return ChallengeAsync(authenticationScheme, properties, ChallengeBehavior.Automatic);
         }
 
-        public virtual void Challenge(string authenticationScheme)
+        public Task SignInAsync(string authenticationScheme, ClaimsPrincipal principal)
         {
-            Challenge(authenticationScheme: authenticationScheme, properties: null);
+            return SignInAsync(authenticationScheme, principal, properties: null);
         }
 
-        public abstract void Challenge(string authenticationScheme, AuthenticationProperties properties);
+        public Task ForbidAsync(string authenticationScheme)
+        {
+            return ForbidAsync(authenticationScheme, properties: null);
+        }
 
-        public void SignIn(string authenticationScheme, ClaimsPrincipal principal)
+        // Deny access (typically a 403)
+        public Task ForbidAsync(string authenticationScheme, AuthenticationProperties properties)
         {
-            SignIn(authenticationScheme, principal, properties: null);
+            return ChallengeAsync(authenticationScheme, properties, ChallengeBehavior.Forbidden);
         }
 
-        public abstract void SignIn(string authenticationScheme, ClaimsPrincipal principal, AuthenticationProperties properties);
+        public abstract Task ChallengeAsync(string authenticationScheme, AuthenticationProperties properties, ChallengeBehavior behavior);
+
+        public abstract Task SignInAsync(string authenticationScheme, ClaimsPrincipal principal, AuthenticationProperties properties);
+
+        public Task SignOutAsync()
+        {
+            return SignOutAsync(authenticationScheme: null, properties: null);
+        }
 
-        public virtual void SignOut()
+        public Task SignOutAsync(string authenticationScheme)
         {
-            SignOut(authenticationScheme: null, properties: null);
+            return SignOutAsync(authenticationScheme, properties: null);
         }
 
-        public abstract void SignOut(string authenticationScheme);
+        public abstract Task SignOutAsync(string authenticationScheme, AuthenticationProperties properties);
 
-        public abstract void SignOut(string authenticationScheme, AuthenticationProperties properties);
+        private static readonly TaskFactory _myTaskFactory = new TaskFactory(CancellationToken.None,
+            TaskCreationOptions.None, TaskContinuationOptions.None, TaskScheduler.Default);
+
+        private static TResult RunSync<TResult>(Func<Task<TResult>> func)
+        {
+            // REVIEW: Do we need to flow culture info still?
+            return _myTaskFactory.StartNew(() => func()).Unwrap().GetAwaiter().GetResult();
```

Comment:
OMFG, what's that? :scream:

It's bad, bad, bad, horribly bad: you're offloading an **async** delegate (`() => func()`) to the thread pool before unwrapping it and then blocking on this new task :smile: 

This clearly creates unnecessary pressure on the thread pool, specially since you use `StartNew` on an `Task`-returning delegate, which is supposed to be non-blocking.